add pagination to elasticsearch results

geraldhumphries opened the pull request:
add pagination to elasticsearch results for each pagination type except 'no'

fix #2387

geraldhumphries commented on generators/client/templates/_bower.json:
```diff
@@ -34,7 +34,7 @@
     "modernizr": "3.2.0",
     "ng-file-upload": "10.0.2",
-    "ngInfiniteScroll": "1.2.1",
+    "ngInfiniteScroll": "1.2.2",
```
This is to fix a bug that was causing only the first one or two pages to load if the results did not fill the whole screen.

geraldhumphries commented:
Changing predicate to `_score` and reverse to `false` means that it will sort by relevance. Clicking on the table headers to sort works.

The separation of `currentSearch` and `searchQuery` is so that the query being searched only changes when clicking the search button, and not when the input changes.

This has been tested for all forms of pagination, with/without services.

geraldhumphries commented:
Bug: when searching with infinite scroll enabled, the sort icon does not get reset.

deepu105 commented on generators/entity/templates/src/main/java/package/service/impl/_EntityServiceImpl.java:
```diff
@@ -96,7 +96,13 @@ public void delete(<%= pkType %> id) {
     @Transactional(readOnly = true) <% } %>
-    public List<<%= instanceType %>> search(String query) {
-        <%- include('../../common/search_template', {viaService: viaService}); -%>
+    public <% if (pagination != 'no') { %>Page<<%= entityClass %><% } else { %>List<<%= instanceType %><% } %>> search(String query<% if (pagination != 'no') { %>, Pageable pageable<% } %>) {
```
Why is the search template ejs not used here?

geraldhumphries replied:
I followed the same pattern as the loadAll service. The template now contains the method signature for the resource search method (like loadAll) so it would not work.

I just noticed the formatting is off however.

deepu105 commented on generators/entity/templates/src/main/java/package/web/rest/_EntityResource.java:
```diff
@@ -137,16 +137,14 @@
-    @Timed
-    public List<<%= instanceType %>> search<%= entityClassPlural %>(@PathVariable String query) {<%- include('../../common/search_template', {viaService: viaService}); -%>
-    }<% } %>
+    @Timed<%- include('../../common/search_template', {viaService: viaService}); -%><% } %>
```
Check if the generated code output is properly formatted plz

geraldhumphries replied:
This part was good, but there were a few other parts with missing/extra spaces.

deepu105 commented on generators/entity/templates/src/main/java/package/service/impl/_EntityServiceImpl.java:
```diff
@@ -95,8 +95,14 @@ public void delete(<%= pkType %> id) {
-    @Transactional(readOnly = true) <% } %>
-    public List<<%= instanceType %>> search(String query) {
-        <%- include('../../common/search_template', {viaService: viaService}); -%>
+    @Transactional(readOnly = true)<% } %>
```
The idea to use templates was to minimize code duplication coz you will do same code when using  a repo directly in resources or when using it in a service and hence the viaService if else in the template.

Isnt the generated code the same for this method and the resource method if used without a service?  If so cant template be resused by passing via service false here as I was doing earlier?

geraldhumphries replied:
I can do that but I'll need to split up search_template. Should we refactor get_all_template too while we're at it?

deepu105 replied:
If it helps maintenance then yes we can split
On 22 Feb 2016 10:02, "Gerald Humphries" <notifications@github.com> wrote:

> In
> generators/entity/templates/src/main/java/package/service/impl/_EntityServiceImpl.java
> <https://github.com/jhipster/generator-jhipster/pull/2953#discussion_r53580321>
> :
>
> > @@ -95,8 +95,14 @@ public void delete(<%= pkType %> id) {
> >       *  @param query the query of the search
> >       *  @return the list of entities
> >       */<% if (databaseType == 'sql') { %>
> > -    @Transactional(readOnly = true) <% } %>
> > -    public List<<%= instanceType %>> search(String query) {
> > -        <%- include('../../common/search_template', {viaService: viaService}); -%>
> > +    @Transactional(readOnly = true)<% } %>
>
> I can do that but I'll need to split up search_template. Should we
> refactor get_all_template too while we're at it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/2953/files#r53580321>
> .
>

geraldhumphries replied:
The stream template was the only one I could find where the code was the same. Is there anywhere else you can see?

geraldhumphries commented:
So, I'm running into this issue regarding the sorting: https://www.elastic.co/guide/en/elasticsearch/guide/master/multi-fields.html

Basically, string fields are analyzed and split up by word. This is critical for searching in these fields, but if there are multiple words in that field then sorting on it will have unexpected results (`ZZZ AAA` will come before `BBB CCC` when sorting alphabetically). This can be solved by adding an unanalyzed "raw" field to the Elasticsearch mapping: (credit to @marquetd for the following)
```
    @MultiField(
            mainField = @Field(type = FieldType.String),
            otherFields = {
                    @NestedField(
                            dotSuffix = "raw",
                            type = FieldType.String,
                            store = true,
                            index = FieldIndex.not_analyzed
                    )
            }
    )
    private String name;
```

Then when sorting, you sort on `field.raw`:
```
sort: $scope.predicate + ($scope.predicate == 'name' ? '.raw' : '')  + ',' + ($scope.ascending ? 'asc' : 'desc')
```

However, in JHipster we can't add that mapping to every string field!

I think we should add this as a tip to the docs after this PR is merged.

deepu105 commented:
can you check the conflicts?

jdubois commented:
I've looked at the code and that's OK for me.
I haven't tested yet as there are some merge conflicts -> @geraldhumphries @deepu105 I see all your discussions, for both of you is this ready to be merged?

geraldhumphries commented:
I'll fix the conflicts and retest everything tonight.

geraldhumphries commented:
There is one bug that I mentioned but I'm not sure how to fix: with infinite scroll enabled, the sort icons don't update when searching. It's managed frombthe sort directive and I don't want to modify the DOM from the controller. Any suggestions? Or should we just open an issue for it and sort it out later?

jdubois commented:
Or we can remove the sort icons with infinite scroll.... It's clear to me that those 2 features are not going to work well together

deepu105 commented:
did you try adding a watch for the parent scope property to reset the icon
from the directive?

Thanks & Regards,
Deepu

On Thu, Feb 25, 2016 at 7:22 AM, Julien Dubois <notifications@github.com>
wrote:

> Or we can remove the sort icons with infinite scroll.... It's clear to me
> that those 2 features are not going to work well together
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/2953#issuecomment-188512221>
> .
>

geraldhumphries commented:
I am going to rebase, then I will try that.

geraldhumphries commented:
Okay, I fixed the sorting issue and rebased down to one commit. The watch worked, and calling applyClass directly from the jh-sort-by directive became redundant so I decided to use watch exclusively to handle icon updates,.
